File: frontend/src/types/cyberark.ts
// =============================================================================
// CyberArk Resource Types
// =============================================================================

export interface CyberArkSafe {
  id: number;
  safe_name: string;
  description: string | null;
  managing_cpm: string | null;
  number_of_members: number;
  number_of_accounts: number;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  is_deleted: boolean;
  updated_at: string;
}

export interface CyberArkSafeMember {
  id: number;
  safe_name: string;
  member_name: string;
  member_type: string;
  permission_level: string | null;
}

export interface CyberArkAccountBrief {
  id: number;
  account_id: string;
  account_name: string;
  platform_id: string | null;
  address: string | null;
  username: string | null;
  secret_type: string | null;
  tf_managed: boolean;
}

export interface CyberArkSafeDetail extends CyberArkSafe {
  members: CyberArkSafeMember[];
  accounts: CyberArkAccountBrief[];
  created_at: string;
}

export interface CyberArkAccount {
  id: number;
  account_id: string;
  account_name: string;
  safe_name: string;
  platform_id: string | null;
  address: string | null;
  username: string | null;
  secret_type: string | null;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  is_deleted: boolean;
  updated_at: string;
}

export interface CyberArkRole {
  id: number;
  role_id: string;
  role_name: string;
  description: string | null;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  is_deleted: boolean;
  updated_at: string;
}

export interface CyberArkRoleMember {
  id: number;
  role_id: string;
  member_name: string;
  member_type: string;
}

export interface CyberArkRoleDetail extends CyberArkRole {
  members: CyberArkRoleMember[];
  created_at: string;
}

export interface CyberArkSIAPolicy {
  id: number;
  policy_id: string;
  policy_name: string;
  policy_type: string;
  description: string | null;
  status: string;
  target_criteria: Record<string, unknown> | null;
  tf_managed: boolean;
  tf_state_source: string | null;
  tf_resource_address: string | null;
  is_deleted: boolean;
  updated_at: string;
}

export interface CyberArkSIAPolicyPrincipal {
  id: number;
  policy_id: string;
  principal_name: string;
  principal_type: string;
}

export interface CyberArkSIAPolicyDetail extends CyberArkSIAPolicy {
  principals: CyberArkSIAPolicyPrincipal[];
  created_at: string;
}

// =============================================================================
// CyberArk Page Types
// =============================================================================

export type CyberArkResourceType = "safes" | "roles" | "sia-policies";

export interface CyberArkFilters {
  search?: string;
  tf_managed?: boolean;
  policy_type?: string;
}

// =============================================================================
// CyberArk List Response Types
// =============================================================================

export interface CyberArkSafeListResponse {
  data: CyberArkSafe[];
  meta: { total: number; last_refreshed: string | null };
}

export interface CyberArkRoleListResponse {
  data: CyberArkRole[];
  meta: { total: number; last_refreshed: string | null };
}

export interface CyberArkSIAPolicyListResponse {
  data: CyberArkSIAPolicy[];
  meta: { total: number; last_refreshed: string | null };
}

// =============================================================================
// Access Mapping Types
// =============================================================================

export interface AccessPathStep {
  entity_type: string;
  entity_id: string;
  entity_name: string;
  context?: Record<string, unknown> | null;
}

export interface AccessPath {
  access_type: "standing" | "jit";
  steps: AccessPathStep[];
}

export interface TargetAccessInfo {
  target_type: "ec2" | "rds";
  target_id: string;
  target_name: string | null;
  target_address: string | null;
  vpc_id: string | null;
  display_status: string | null;
  instance_type: string | null;
  engine: string | null;
  platform: string | null;
  access_paths: AccessPath[];
}

export interface UserAccessMapping {
  user_name: string;
  targets: TargetAccessInfo[];
  access_paths?: AccessPath[];
}

export interface AccessMappingResponse {
  users: UserAccessMapping[];
  total_users: number;
  total_targets: number;
  total_standing_paths: number;
  total_jit_paths: number;
}

export interface AccessMappingUserList {
  users: string[];
}

export interface AccessMappingTargetBrief {
  target_type: string;
  target_id: string;
  target_name: string | null;
  target_address: string | null;
}

export interface AccessMappingTargetList {
  targets: AccessMappingTargetBrief[];
}

// =============================================================================
// Access Mapping Node Data (Collapsible)
// =============================================================================

export interface AccessNodeChildSummary {
  roleCount?: number;
  safeCount?: number;
  policyCount?: number;
  accountCount?: number;
  targetCount?: number;
  standingCount?: number;
  jitCount?: number;
}

export interface CollapsibleNodeData {
  collapsed?: boolean;
  onToggleCollapse?: () => void;
  childSummary?: AccessNodeChildSummary;
}

// =============================================================================
// CyberArk User Types (SCIM)
// =============================================================================

export interface CyberArkIdentityUser {
  id: number;
  user_id: string;
  user_name: string;
  display_name: string | null;
  email: string | null;
  active: boolean;
  is_deleted: boolean;
  updated_at: string;
}

export interface CyberArkUserListResponse {
  data: CyberArkIdentityUser[];
  meta: { total: number; last_refreshed: string | null };
}

// =============================================================================
// CyberArk Settings Types
// =============================================================================

export interface CyberArkSettingsResponse {
  tenant_name: string | null;
  enabled: boolean;
  base_url: string | null;
  identity_url: string | null;
  uap_base_url: string | null;
  client_id: string | null;
  has_client_secret: boolean;
  updated_at: string | null;
  updated_by: string | null;
}

export interface CyberArkSettingsUpdate {
  tenant_name?: string;
  enabled?: boolean;
  base_url?: string;
  identity_url?: string;
  uap_base_url?: string;
  client_id?: string;
  client_secret?: string;
}

export interface CyberArkConnectionTestRequest {
  base_url: string;
  identity_url: string;
  client_id: string;
  client_secret?: string;
}

export interface TenantDiscoveryRequest {
  subdomain: string;
}

export interface TenantDiscoveryResponse {
  success: boolean;
  base_url: string | null;
  identity_url: string | null;
  uap_base_url: string | null;
  region: string | null;
  message: string | null;
}

export interface CyberArkConnectionTestResponse {
  success: boolean;
  message: string;
  details?: Record<string, unknown>;
}

export interface CyberArkSyncStatus {
  config: {
    source: string;
    enabled: boolean;
    all_fields_set: boolean;
    db_settings_exists: boolean;
    db_enabled: boolean | null;
    db_base_url_set: boolean;
    db_identity_url_set: boolean;
    db_client_id_set: boolean;
    db_client_secret_set: boolean;
  };
  database_counts: {
    roles_total: number;
    roles_active: number;
    safes: number;
    accounts: number;
    sia_policies: number;
    users: number;
  };
  last_sync: {
    synced_at: string | null;
    status: string | null;
    resource_count: number | null;
  };
}

// =============================================================================
// SCIM Settings Types
// =============================================================================

export interface ScimSettingsResponse {
  scim_enabled: boolean;
  scim_app_id: string | null;
  scim_oauth2_url: string | null;
  scim_scope: string | null;
  scim_client_id: string | null;
  has_scim_client_secret: boolean;
  updated_at: string | null;
  updated_by: string | null;
}

export interface ScimSettingsUpdate {
  scim_enabled?: boolean;
  scim_app_id?: string;
  scim_scope?: string;
  scim_client_id?: string;
  scim_client_secret?: string;
}

export interface ScimConnectionTestRequest {
  scim_app_id?: string;
  scim_oauth2_url?: string;
  scim_scope: string;
  scim_client_id: string;
  scim_client_secret?: string;
}

export interface ScimConnectionTestResponse {
  success: boolean;
  message: string;
  details?: Record<string, unknown>;
}

// =============================================================================
// CyberArk Drift Types
// =============================================================================

export interface CyberArkDriftResponse {
  drift_detected: boolean;
  items: Array<{
    resource_type: string;
    resource_id: string;
    drift_type: "unmanaged" | "orphaned" | "modified";
    details: string | null;
  }>;
  checked_at: string;
}

File: backend/app/schemas/cyberark.py
"""
Pydantic schemas for CyberArk resources and access mapping.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict


class BaseSchema(BaseModel):
    """Base schema with common configuration."""

    model_config = ConfigDict(from_attributes=True)


# =============================================================================
# CyberArk Safe Schemas
# =============================================================================


class CyberArkSafeMemberResponse(BaseSchema):
    """Safe member response schema."""

    id: int
    safe_name: str
    member_name: str
    member_type: str  # user, group, role
    permission_level: Optional[str] = None


class CyberArkAccountBrief(BaseSchema):
    """Brief account info for safe detail views."""

    id: int
    account_id: str
    account_name: str
    platform_id: Optional[str] = None
    address: Optional[str] = None
    username: Optional[str] = None
    secret_type: Optional[str] = None
    tf_managed: bool = False


class CyberArkSafeResponse(BaseSchema):
    """CyberArk safe response schema."""

    id: int
    safe_name: str
    description: Optional[str] = None
    managing_cpm: Optional[str] = None
    number_of_members: int = 0
    number_of_accounts: int = 0
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    is_deleted: bool = False
    updated_at: datetime


class CyberArkSafeDetail(CyberArkSafeResponse):
    """Detailed safe response with members and accounts."""

    members: List[CyberArkSafeMemberResponse] = []
    accounts: List[CyberArkAccountBrief] = []
    created_at: datetime


# =============================================================================
# CyberArk Account Schemas
# =============================================================================


class CyberArkAccountResponse(BaseSchema):
    """CyberArk account response schema."""

    id: int
    account_id: str
    account_name: str
    safe_name: str
    platform_id: Optional[str] = None
    address: Optional[str] = None
    username: Optional[str] = None
    secret_type: Optional[str] = None
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    is_deleted: bool = False
    updated_at: datetime


# =============================================================================
# CyberArk Role Schemas
# =============================================================================


class CyberArkRoleMemberResponse(BaseSchema):
    """Role member response schema."""

    id: int
    role_id: str
    member_name: str
    member_type: str  # user, group


class CyberArkRoleResponse(BaseSchema):
    """CyberArk role response schema."""

    id: int
    role_id: str
    role_name: str
    description: Optional[str] = None
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    is_deleted: bool = False
    updated_at: datetime


class CyberArkRoleDetail(CyberArkRoleResponse):
    """Detailed role response with members."""

    members: List[CyberArkRoleMemberResponse] = []
    created_at: datetime


# =============================================================================
# CyberArk SIA Policy Schemas
# =============================================================================


class CyberArkSIAPolicyPrincipalResponse(BaseSchema):
    """SIA policy principal response schema."""

    id: int
    policy_id: str
    principal_name: str
    principal_type: str  # user, role


class CyberArkSIAPolicyResponse(BaseSchema):
    """CyberArk SIA policy response schema."""

    id: int
    policy_id: str
    policy_name: str
    policy_type: str  # vm, database
    description: Optional[str] = None
    status: str
    target_criteria: Optional[Dict[str, Any]] = None
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    is_deleted: bool = False
    updated_at: datetime


class CyberArkSIAPolicyDetail(CyberArkSIAPolicyResponse):
    """Detailed SIA policy response with principals."""

    principals: List[CyberArkSIAPolicyPrincipalResponse] = []
    created_at: datetime


# =============================================================================
# Access Mapping Schemas
# =============================================================================


class AccessPathStep(BaseSchema):
    """A single step in an access path."""

    entity_type: str  # user, role, safe, account, sia_policy
    entity_id: str
    entity_name: str
    context: Optional[Dict[str, Any]] = None


class AccessPath(BaseSchema):
    """A complete access path from user to target."""

    access_type: str  # standing, jit
    steps: List[AccessPathStep]


class TargetAccessInfo(BaseSchema):
    """Information about a target and the access paths to it."""

    target_type: str  # ec2, rds
    target_id: str
    target_name: Optional[str] = None
    target_address: Optional[str] = None
    vpc_id: Optional[str] = None
    display_status: Optional[str] = None
    instance_type: Optional[str] = None  # EC2 instance type or RDS instance class
    engine: Optional[str] = None  # RDS engine (e.g., "postgres", "mysql")
    platform: Optional[str] = None  # Platform label (e.g., "Linux", "PostgreSQL")
    tf_managed: bool = False
    access_paths: List[AccessPath] = []


class UserAccessMapping(BaseSchema):
    """All targets accessible by a specific user."""

    user_name: str
    targets: List[TargetAccessInfo] = []
    access_paths: List[AccessPath] = []  # Relationship-only paths (no target)


class AccessMappingResponse(BaseSchema):
    """Response for the access mapping visualization endpoint."""

    users: List[UserAccessMapping] = []
    total_users: int = 0
    total_targets: int = 0
    total_standing_paths: int = 0
    total_jit_paths: int = 0


class AccessMappingUserList(BaseSchema):
    """List of users available for access mapping."""

    users: List[str] = []


class AccessMappingTargetBrief(BaseSchema):
    """Brief target info for the access mapping filter dropdown."""

    target_type: str
    target_id: str
    target_name: Optional[str] = None
    target_address: Optional[str] = None


class AccessMappingTargetList(BaseSchema):
    """List of targets available for access mapping."""

    targets: List[AccessMappingTargetBrief] = []


# =============================================================================
# CyberArk Settings Schemas
# =============================================================================


class CyberArkUserResponse(BaseSchema):
    """CyberArk Identity user response schema."""

    id: int
    user_id: str
    user_name: str
    display_name: Optional[str] = None
    email: Optional[str] = None
    active: bool = True
    tf_managed: bool = False
    tf_state_source: Optional[str] = None
    tf_resource_address: Optional[str] = None
    is_deleted: bool = False
    updated_at: datetime


class CyberArkUserListResponse(BaseSchema):
    """CyberArk user list response."""

    data: List[CyberArkUserResponse] = []
    meta: Dict[str, Any] = {}


# =============================================================================
# CyberArk Settings Schemas
# =============================================================================


class CyberArkSettingsResponse(BaseSchema):
    """CyberArk settings response."""

    tenant_name: Optional[str] = None
    enabled: bool = False
    base_url: Optional[str] = None
    identity_url: Optional[str] = None
    uap_base_url: Optional[str] = None
    client_id: Optional[str] = None
    has_client_secret: bool = False
    updated_at: Optional[datetime] = None
    updated_by: Optional[str] = None


class CyberArkSettingsUpdate(BaseSchema):
    """CyberArk settings update request."""

    tenant_name: Optional[str] = None
    enabled: Optional[bool] = None
    base_url: Optional[str] = None
    identity_url: Optional[str] = None
    uap_base_url: Optional[str] = None
    client_id: Optional[str] = None
    client_secret: Optional[str] = None


class CyberArkConnectionTestRequest(BaseSchema):
    """Request to test CyberArk API connection."""

    base_url: str
    identity_url: str
    client_id: str
    client_secret: Optional[str] = None


class TenantDiscoveryRequest(BaseSchema):
    """Request to discover CyberArk tenant URLs."""

    subdomain: str


class TenantDiscoveryResponse(BaseSchema):
    """Response from CyberArk tenant discovery."""

    success: bool
    base_url: Optional[str] = None
    identity_url: Optional[str] = None
    uap_base_url: Optional[str] = None
    region: Optional[str] = None
    message: Optional[str] = None


class CyberArkConnectionTestResponse(BaseSchema):
    """Response from CyberArk connection test."""

    success: bool
    message: str
    details: Optional[Dict[str, Any]] = None


# =============================================================================
# SCIM Settings Schemas
# =============================================================================


class ScimSettingsResponse(BaseSchema):
    """SCIM integration settings response."""

    scim_enabled: bool = False
    scim_app_id: Optional[str] = None
    scim_oauth2_url: Optional[str] = None
    scim_scope: Optional[str] = None
    scim_client_id: Optional[str] = None
    has_scim_client_secret: bool = False
    updated_at: Optional[datetime] = None
    updated_by: Optional[str] = None


class ScimSettingsUpdate(BaseSchema):
    """SCIM settings update request."""

    scim_enabled: Optional[bool] = None
    scim_app_id: Optional[str] = None
    scim_scope: Optional[str] = None
    scim_client_id: Optional[str] = None
    scim_client_secret: Optional[str] = None


class ScimConnectionTestRequest(BaseSchema):
    """Request to test SCIM OAuth2 connection."""

    scim_app_id: Optional[str] = None
    scim_oauth2_url: Optional[str] = None
    scim_scope: str
    scim_client_id: str
    scim_client_secret: Optional[str] = None


class ScimConnectionTestResponse(BaseSchema):
    """Response from SCIM connection test."""

    success: bool
    message: str
    details: Optional[Dict[str, Any]] = None
